### PR TITLE
Hookmap.run does not swallow tracebacks

### LINKED ISSUE #1770: HookMap.run swallows tracebacks
**I'm submitting a ...**
- [x] bug report
- [ ] feature request
- [ ] question about the decisions made in the repository

**Do you want to request a *feature* or report a *bug*?**

Report a bug.

**What is the current behavior?**

`HookMap.run` does not preserve traceback information when passing an exception.

If the exception is not `HTTPError`, `HTTPRedirect`, `InternalRedirect`, `KeyboardInterrupt` or `SystemExit`, it is logged with traceback through `cherrypy.log`, but then the exception is re-raised without preserving the original traceback.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a screenshots and logs of the problem. If you can, show us your code.**

I could provide all that if that's really needed, but come on...

**What is the expected behavior?**

Re-raise exceptions from hooks in a way that preserves the original traceback.

**What is the motivation / use case for changing the behavior?**

Making it easier to debug tools from automated tests using `webtest`. In my environment, logging is suppressed to reduce noise, but then tracebacks originating from tools are lost.

**Please tell us about your environment:**

- Cheroot version: 6.5.4
- CherryPy version: 17.4.1
- Python version: 2.7.15
- OS: Not relevant
- Browser: Not relevant

**Other information**

The fix is trivial,
https://github.com/cherrypy/cherrypy/blob/master/cherrypy/_cprequest.py#L114
Replace with a bare `raise` statement.

The bare `raise` statement was introduced in Python 1.5.1p1… It provides exactly the semantics wanted here.

https://docs.python.org/release/1.5.1p1/ref/raise.html
https://docs.python.org/release/2.7.15/reference/simple_stmts.html#the-raise-statement
https://docs.python.org/release/3.7.2/reference/simple_stmts.html#the-raise-statement

PR upcoming.

### PR DESCRIPTION
**Pull request by ddaanet**
**What kind of change does this PR introduce?**
  - [x] bug fix
  - [ ] feature
  - [ ] docs update
  - [ ] tests/coverage improvement
  - [ ] refactoring
  - [ ] other

**What is the related issue number (starting with `#`)**

Fixes #1770

**What is the current behavior?** (You can also link to an open issue here)

See issue.

**What is the new behavior (if this is a feature change)?**

See issue.

**Other information**:

Would be nice to have it backported in 17.

**Checklist**:

  - [x] I think the code is well written
  - [x] I wrote [good commit messages][1]
  - [ ] I have [squashed related commits together][2] after the changes have been approved
  - [ ] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [ ] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

### CODE REVIEW

**Comment by ddaanet** [comment]
Really, this is a lot of hoops jumping for such a simple fix, but this one has been itching me one time too many.

File: cherrypy/_cprequest.py [hunk 1]
```diff
@@ -99,19 +99,19 @@ def run(self, point):
...
-                    exc = sys.exc_info()[1]
+                    raising = True
                 except Exception:
-                    exc = sys.exc_info()[1]
+                    raising = True
                     cherrypy.log(traceback=True, severity=40)
-        if exc:
-            raise exc
+        if raising:
+            raise
```
Comment (webknjaz) [inline cherrypy/_cprequest.py:114]: I'm pretty sure this will not work because there's nothing to raise in this context (outside of the except block)...

**Comment by webknjaz** [comment]
Is this covered by a known test? Can you provide a minimal reproducer demonstrating what you mean to a stranger?

**Comment by ddaanet** [comment]
Ok. This is getting harder than I expected. Apparrently that relies on a poorly documented behavior of Python. I will work on it.

**Comment by webknjaz** [comment]
What's poorly documented? In Python 3 it should be working just as you want due to automatic exception chaining and storing `__cause__` or/and `__context__` as per https://www.python.org/dev/peps/pep-3134/

**Comment by jaraco** [comment]
It would be particularly nice if this solution could emit _all_ tracebacks encountered during the hook. In other words:

```python
def run(self):
  run_hooks(iter(sorted(self[point])))

def run_hooks(hooks):
  for hook in hooks:
    try:
      hook.run()
    except Exception:
      run_hooks(filter(operator.attrgetter('failsafe'), hooks))
      raise
```

**Comment by webknjaz** [comment]
@jaraco yeah, it'd be nice. I know that some of asyncio devs were thinking about adding another attribute for multiple exception causes to the base exception. While this is useful in the async world in particular, you've correctly pointed out that it's applicable to our architecture as well...

**Comment by jaraco** [comment]
Thanks to webknjaz's reference and looking at this problem while at the CPython sprints, Lisa Roach was able to assist with a couple of references of existing art around handling multiple exceptions in an async model:

- [edb MultiError](https://github.com/edgedb/edgedb/blob/4623b0bc0172999d26357e0a4b9b6bb47a45b47d/edb/common/taskgroup.py#L267-L286)
- [trio MultiError](https://github.com/python-trio/trio/blob/c47c08cbe65790e5f0dfef15571a17b727dbc338/trio/_core/_multierror.py)

Perhaps one of these could be re-used or at least used for inspiration.

Although in our case, we're operating within a single thread synchronously, so maybe the suggestion I've made would be best - rely on the default multiple exception handling behavior in Python to catch the exceptions in the hooks.

Regardless how this is handled, this problem needs a different approach than the one proposed.

**Pull request closed without merging**